compiler/rustc_session/src/session.rs

```diff
@@ -1302,10 +1302,24 @@ fn validate_commandline_args_with_session_available(sess: &Session) {
             .emit_err(errors::SplitDebugInfoUnstablePlatform { debuginfo: sess.split_debuginfo() });
     }
 
-    let dwarf_version =
-        sess.opts.unstable_opts.dwarf_version.unwrap_or(sess.target.default_dwarf_version);
-    if sess.opts.unstable_opts.embed_source && dwarf_version < 5 {
-        sess.dcx().emit_err(errors::EmbedSourceInsufficientDwarfVersion { dwarf_version });
+    if sess.opts.unstable_opts.embed_source {
+        let dwarf_version =
+            sess.opts.unstable_opts.dwarf_version.unwrap_or(sess.target.default_dwarf_version);
+
+        let uses_llvm_backend =
+            matches!(sess.opts.unstable_opts.codegen_backend.as_deref(), None | Some("llvm"));
+
+        if dwarf_version < 5 {
+            sess.dcx().emit_err(errors::EmbedSourceInsufficientDwarfVersion { dwarf_version });
+        }
+
+        if sess.opts.debuginfo == DebugInfo::None {
+            sess.dcx().emit_err(errors::EmbedSourceRequiresDebugInfo);
+        }
+
+        if !uses_llvm_backend {
+            sess.dcx().emit_err(errors::EmbedSourceRequiresLLVMBackend);
```

Comment:
Do we error like this for other features that only support some backends? Like split debuginfo?